sql: panic when formatting Tuple type

**Describe the problem**

The latest version of the CockroachDB (b861696207) `panics` when executing the following query:

```sql
create table dummy_table(v1 int);
SELECT * FROM dummy_table WHERE ((999999999999999999, 0, 0) AS required_bug_triggering_alias);
```

**To Reproduce**

Here is the detail steps to reproduce the bug.

1. In operating system Ubuntu 20.04, download the CockroachDB source code from the github source.
2. Use the latest version of the CockroachDB code (tested version: b861696207)
3. Directly `make install` in the root repository folder. 
4. Run `./cockroach demo`, and then paste the PoC query to the cockroach cli environment.
5. Observe the `panic` and log the stack information. 

**Expected behavior**
The `SELECT` statement should return semantic error as the `WHERE` clause does not return a boolean data type. 

**Additional data / screenshots**

Here is the outputted stack trace: 

```
demo@127.0.0.1:26257/movr> create table dummy_table(v1 int);
CREATE TABLE

Time: 16ms total (execution 13ms / network 3ms)

demo@127.0.0.1:26257/movr> SELECT * FROM dummy_table WHERE ((999999999999999999, 0, 0) AS required_bug_triggering_alias);
*
* ERROR: a SQL panic has occurred while executing the following statement:
* SELECT * FROM dummy_table WHERE ((999999999999999999, 0, 0) AS required_bug_triggering_alias)
*
panic: runtime error: slice bounds out of range [:2] with capacity 1 [recovered]
	panic: runtime error: slice bounds out of range [:2] with capacity 1

goroutine 2643 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Tuple).formatHideConstants(0x4004784780, 0x40019d5340?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:129 +0x338
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0x40019d5340, {0x82791c0?, 0x4004784780?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:30 +0xe4
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x40019d5340, {0x82791c0, 0x4004784780})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:452 +0x94
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Where).Format(0x4004bb55e0, 0x40019d5340)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/select.go:745 +0x70
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0x40019d5340?, {0x82794e0?, 0x4004bb55e0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:49 +0x1c8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x40019d5340, {0x82794e0, 0x4004bb55e0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:452 +0x94
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*SelectClause).Format(0x40019d75e0, 0x40019d5340)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/select.go:125 +0x15c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0x40019d5340?, {0x8278640?, 0x40019d75e0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:49 +0x1c8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x40019d5340, {0x8278640, 0x40019d75e0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:452 +0x94
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Select).Format(0x4004b0d2c0, 0x40019d5340)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/select.go:57 +0x60
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0x40019d5340?, {0x8278620?, 0x4004b0d2c0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:49 +0x1c8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x40019d5340, {0x8278620, 0x4004b0d2c0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:452 +0x94
github.com/cockroachdb/cockroach/pkg/sql.anonymizeStmtAndConstants({0x82e0d80?, 0x4004b0d2c0}, {0x82b84b8?, 0x4000bf3a10?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:2276 +0xac
github.com/cockroachdb/cockroach/pkg/sql.WithAnonymizedStatement({0x827a200, 0x400488a8e8}, {0x82e0d80?, 0x4004b0d2c0?}, {0x82b84b8?, 0x4000bf3a10?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:2283 +0x44
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0x40024c6000, {0x82b7450, 0x4002deb280}, {0x4b1d3a0?, 0x4004a72fa8?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1072 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x82b7450, 0x4002deb280}, {0x4?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:834 +0x4c
panic({0x4b1d3a0, 0x4004a72fa8})
	/usr/local/go/src/runtime/panic.go:884 +0x20c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Tuple).formatHideConstants(0x4004784780, 0x4002b76a50?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:129 +0x338
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0x4002b76a50, {0x82791c0?, 0x4004784780?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:30 +0xe4
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x4002b76a50, {0x82791c0, 0x4004784780})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:452 +0x94
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Where).Format(0x4004bb55e0, 0x4002b76a50)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/select.go:745 +0x70
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0x4002b76a50?, {0x82794e0?, 0x4004bb55e0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:49 +0x1c8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x4002b76a50, {0x82794e0, 0x4004bb55e0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:452 +0x94
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*SelectClause).Format(0x40019d75e0, 0x4002b76a50)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/select.go:125 +0x15c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0x4002b76a50?, {0x8278640?, 0x40019d75e0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:49 +0x1c8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x4002b76a50, {0x8278640, 0x40019d75e0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:452 +0x94
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Select).Format(0x4004b0d2c0, 0x4002b76a50)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/select.go:57 +0x60
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0x4002b76a50?, {0x8278620?, 0x4004b0d2c0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:49 +0x1c8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x4002b76a50, {0x8278620, 0x4004b0d2c0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:452 +0x94
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.AsStringWithFlags({0x8278620, 0x4004b0d2c0}, 0x100, {0x0?, 0x173c1be1233beb99?, 0xc257c0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:590 +0x64
github.com/cockroachdb/cockroach/pkg/sql.formatStatementHideConstants(...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:3516
github.com/cockroachdb/cockroach/pkg/sql.makeStatement({{0x82e0d80, 0x4004b0d2c0}, {0x0, 0x0, 0x0}, {0x4002401299, 0x5d}, 0x0, 0x1}, {{0x0?, ...}})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/statement.go:45 +0xa0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0x40024c6000, {0x82b74f8?, 0x400380e0f0?}, {{0x82e0d80, 0x4004b0d2c0}, {0x0, 0x0, 0x0}, {0x4002401299, 0x5d}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:285 +0x308
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x82b74f8?, 0x400380e0f0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130 +0x8c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x82b74f8?, {0x82b74f8?, 0x400380e0f0?}, {0x82e0d80?, 0x4004b0d2c0?}, 0x0?, 0xaf37fe0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2527 +0x280
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0x40024c6000, {0x82b74f8, 0x400380e0f0}, {{0x82e0d80, 0x4004b0d2c0}, {0x0, 0x0, 0x0}, {0x4002401299, 0x5d}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129 +0x388
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0x4004b41ab8, 0x40024c6000, 0x4004b417b0, 0x4004b41868, 0x4004b418a8, 0x4004b41888, 0x4004b41878, 0x4004b41898)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1959 +0x2dc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0x40024c6000)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1964 +0xcdc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0x40024c6000, {0x82b7450, 0x4002deb280}, 0xf673c8?, 0x40024b5dd8?, 0x40032d63d0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1882 +0x18c
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0x4001f9aaa0?, {0x82b7450?, 0x4002deb280?}, {0x4002deb280?}, 0x8305868?, 0x4002134990?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:836 +0xb4
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:731 +0x290
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:644 +0x1e8
```

**Environment:**

 - CockroachDB version: v23.1.0-alpha
 - Server OS: Ubuntu 20.04 LTS
 - Client app: CockroachDB demo command. (`./cockroach demo`)
 - Detailed CockroachDB version output:

```
cockroach version details:
Build Tag:        v23.1.0-alpha.1-1132-gb861696207
Build Time:       2023/01/20 18:12:24
Distribution:     CCL
Platform:         linux arm64 (aarch64-linux-gnu)
Go Version:       go1.19
C Compiler:       gcc 9.4.0
Build Commit ID:  b861696207d6640f187add625ac6e4bb7b9575d4
Build Type:       development
```

**Additional context**

The bug can be reproduced by CockroachDB release version:  `CockroachDB v23.1 Testing Releases`  , `v22.2.2` and `v22.1`. We haven't tested CockroachDB release versions before `v22.1`.  


Jira issue: CRDB-23625

seem like we do not need a very large number to repro this
```
demo@127.0.0.1:26257/movr> SELECT * FROM dummy_table WHERE ((0, 0, 0) AS required_bug_triggering_alias);
*
* ERROR: [n1,client=127.0.0.1:59161,hostssl,user=demo] a SQL panic has occurred while executing the following statement:
* SELECT * FROM dummy_table WHERE ((0, 0, 0) AS required_bug_triggering_alias)
*
panic: runtime error: slice bounds out of range [:2] with capacity 1 [recovered]
	panic: runtime error: slice bounds out of range [:2] with capacity 1

goroutine 6085 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Tuple).formatHideConstants(0xc00252efc0, 0xc0039b1140)
```

--- a/pkg/sql/parser/testdata/select_exprs
+++ b/pkg/sql/parser/testdata/select_exprs
@@ -233,6 +233,14 @@ SELECT ((_,) AS a) -- literals removed
 SELECT ((1,) AS _) -- identifiers removed
 
 parse
+SELECT (ROW(1) AS a, b)
+----
+SELECT ((1,) AS a, b) -- normalized!
+SELECT ((((1),) AS a, b)) -- fully parenthesized
+SELECT ((_,) AS a, b) -- literals removed
+SELECT ((1,) AS _, _) -- identifiers removed
+
+parse
 SELECT (SELECT 1)
 ----
 SELECT (SELECT 1)
@@ -311,6 +319,38 @@ SELECT ((1, 2, 3) AS a, b, c)
 SELECT ((((1), (2), (3)) AS a, b, c)) -- fully parenthesized
 SELECT ((_, _, __more1_10__) AS a, b) -- literals removed
 SELECT ((1, 2, 3) AS _, _, _) -- identifiers removed
+
+parse
+SELECT ((1, 2, 3) AS a)
+----
+SELECT ((1, 2, 3) AS a)
+SELECT ((((1), (2), (3)) AS a)) -- fully parenthesized
+SELECT ((_, _, __more1_10__) AS a) -- literals removed
+SELECT ((1, 2, 3) AS _) -- identifiers removed
+
+parse
+SELECT ((1, 2, 3) AS a, b)
+----
+SELECT ((1, 2, 3) AS a, b)
+SELECT ((((1), (2), (3)) AS a, b)) -- fully parenthesized
+SELECT ((_, _, __more1_10__) AS a, b) -- literals removed
+SELECT ((1, 2, 3) AS _, _) -- identifiers removed
+
+parse
+SELECT ((1, 2) AS a, b, c)
+----
+SELECT ((1, 2) AS a, b, c)
+SELECT ((((1), (2)) AS a, b, c)) -- fully parenthesized
+SELECT ((_, _) AS a, b, c) -- literals removed
+SELECT ((1, 2) AS _, _, _) -- identifiers removed
+
+parse
+SELECT ((1, 2, 3) AS a, b, c, d)
+----
+SELECT ((1, 2, 3) AS a, b, c, d)
+SELECT ((((1), (2), (3)) AS a, b, c, d)) -- fully parenthesized
+SELECT ((_, _, __more1_10__) AS a, b) -- literals removed
+SELECT ((1, 2, 3) AS _, _, _, _) -- identifiers removed
 
 parse
 SELECT ((1, 2, 3))

--- a/pkg/sql/sem/tree/format_test.go
+++ b/pkg/sql/sem/tree/format_test.go
@@ -253,6 +253,46 @@ func TestFormatExpr(t *testing.T) {
 	}
 }
 
+// TestFormatUntypedExpr is similar to TestFormatExpr, but it does not
+// type-check the test expressions before formatting them. It allows for testing
+// formatting of mis-typed expressions.
+func TestFormatUntypedExpr(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	testData := []struct {
+		expr     string
+		f        tree.FmtFlags
+		expected string
+	}{
+		{"((1, 2) AS foo)", tree.FmtHideConstants, "((_, _) AS foo)"},
+		{"((1, 2, 3) AS foo)", tree.FmtHideConstants, "((_, _, __more1_10__) AS foo)"},
+		{"((1, 2, 3) AS foo, bar)", tree.FmtHideConstants, "((_, _, __more1_10__) AS foo, bar)"},
+		{"((1, 2, 3) AS foo, bar, baz)", tree.FmtHideConstants, "((_, _, __more1_10__) AS foo, bar)"},
+		{"((1, 2) AS foo, bar)", tree.FmtHideConstants, "((_, _) AS foo, bar)"},
+		{"((1, 2) AS foo, bar, baz)", tree.FmtHideConstants, "((_, _) AS foo, bar, baz)"},
+		{"(ROW(1) AS foo)", tree.FmtHideConstants, "((_,) AS foo)"},
+		{"(ROW(1) AS foo, bar)", tree.FmtHideConstants, "((_,) AS foo, bar)"},
+		{"(ROW(1) AS foo, bar, baz)", tree.FmtHideConstants, "((_,) AS foo, bar, baz)"},
+		{"(ROW(1, 2) AS foo, bar, baz)", tree.FmtHideConstants, "((_, _) AS foo, bar, baz)"},
+		{"(ROW(1, 2, 3) AS foo, bar)", tree.FmtHideConstants, "((_, _, __more1_10__) AS foo, bar)"},
+		{"(ROW(1, 2, 3) AS foo, bar, baz)", tree.FmtHideConstants, "((_, _, __more1_10__) AS foo, bar)"},
+		{"(ROW(1, 2, 3) AS foo)", tree.FmtHideConstants, "((_, _, __more1_10__) AS foo)"},
+	}
+
+	for i, test := range testData {
+		t.Run(fmt.Sprintf("%d %s", i, test.expr), func(t *testing.T) {
+			expr, err := parser.ParseExpr(test.expr)
+			if err != nil {
+				t.Fatal(err)
+			}
+			exprStr := tree.AsStringWithFlags(expr, test.f)
+			if exprStr != test.expected {
+				t.Fatalf("expected %q, got %q", test.expected, exprStr)
+			}
+		})
+	}
+}
+
 func TestFormatExpr2(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)

--- a/pkg/sql/sem/tree/hide_constants.go
+++ b/pkg/sql/sem/tree/hide_constants.go
@@ -125,7 +125,7 @@ func (node *Tuple) formatHideConstants(ctx *FmtCtx) {
 		v2.Exprs = append(make(Exprs, 0, 3), v2.Exprs[:2]...)
 		if len(node.Exprs) > 2 {
 			v2.Exprs = append(v2.Exprs, arityIndicator(len(node.Exprs)-2))
-			if node.Labels != nil {
+			if len(node.Labels) > 2 {
 				v2.Labels = node.Labels[:2]
 			}
 		}

